### PR TITLE
Allow the file pattern used in LocalIndexRepo to be customised

### PR DESCRIPTION
**Pull request by eli-jordan**
Some bundles don't use the standard naming convention for osgi bundles, which prevents them from being indexed by the LocalIndexedRepo. This pull request allows the pattern to be customised.

### CODE REVIEW

File: biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java [hunk 1]
```diff
@@ -1,46 +1,82 @@
...
+	public static final String	PROP_OVERWRITE				= "overwrite";
+	public static final String	PROP_ONLYDIRS				= "onlydirs";
+	public static final String	PROP_FILE_INCLUDE_POLICY	= "file.include.policy";
+
+	static final String	VERSIONED_BUNDLE_PATTERN	= "([-a-zA-z0-9_\\.]+)-([0-9\\.]+)(-[-a-zA-z0-9_]+)?\\.(jar|lib)";
+	static final String	ALL_JARS_PATTERN			= ".*jar$";
+	static final String	ALL_JARS_AND_LIBS_PATTERN	= ".*(jar|lib)$";
+
+	public enum FileIncludePolicy {
+		VERSIONED_BUNDLE(VERSIONED_BUNDLE_PATTERN), ALL_JARS(ALL_JARS_PATTERN), ALL_JARS_AND_LIBS(
```
Comment (bjhargrave) [inline biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java]: These capitalized enum names are not very nice to use in a bnd file. I think better names are needed.

`...;file.include-policy=ALL_JARS` looks awkward.

File: biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java [hunk 2]
```diff
@@ -1,46 +1,82 @@
...
 public class LocalIndexedRepo extends FixedIndexedRepo implements Refreshable, Participant, Actionable {
+	private final String		UPWARDS_ARROW				= " \u2191";
+	private final String		DOWNWARDS_ARROW				= " \u2193";
+	private static final String	CACHE_PATH					= ".cache";
+	public static final String	PROP_LOCAL_DIR				= "local";
+	public static final String	PROP_READONLY				= "readonly";
+	public static final String	PROP_PRETTY					= "pretty";
+	public static final String	PROP_OVERWRITE				= "overwrite";
+	public static final String	PROP_ONLYDIRS				= "onlydirs";
+	public static final String	PROP_FILE_INCLUDE_POLICY	= "file.include.policy";
```
Comment (bjhargrave) [inline biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java]: The prop names should be shorter and no dots like the other props. Perhaps just `files`?

File: biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java [hunk 3]
```diff
@@ -69,6 +105,30 @@ public synchronized void setProperties(Map<String,String> map) {
 			throw new IllegalArgumentException(String.format(
 					"Cannot create repository cache: '%s' already exists but is not directory.",
 					cacheDir.getAbsolutePath()));
+
+		setIncludePolicy(map);
```
Comment (bjhargrave) [inline biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java]: Why pass the whole map when only the files prop is needed? Just pass the prop value.

File: biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java [hunk 4]
```diff
@@ -69,6 +105,30 @@ public synchronized void setProperties(Map<String,String> map) {
...
+		String policy = props.get(PROP_FILE_INCLUDE_POLICY);
+		if (policy != null && !policy.isEmpty()) {
+			try {
+				FileIncludePolicy spec = FileIncludePolicy.valueOf(policy);
+				includePolicy = spec.pattern;
+			}
+			// this means the policy was not recognized, so try to interpret it
+			// as a regex
+			catch (IllegalArgumentException e) {
+				includePolicy = Pattern.compile(policy);
```
Comment (bjhargrave) [inline biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java]: You must handle the `PatternSyntaxException` here. You can't just let it bubble up. An error must be reported as the statement is malformed.

File: biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java [hunk 5]
```diff
@@ -100,8 +161,14 @@ public synchronized void setProperties(Map<String,String> map) {
 	}
 
 	/**
-	 * @param contentProvider the repository content provider
-	 * @return the filename of the index on local storage
+	 * <<<<<<< HEAD
```
Comment (bjhargrave) [inline biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java]: This is left overs from a merge error.

File: biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java [hunk 6]
```diff
@@ -168,7 +236,7 @@ private synchronized void generateIndex(File indexFile, IRepositoryContentProvid
 	}
 
 	@SuppressWarnings("deprecation")
-	private void gatherFiles(Set<File> allFiles) throws Exception {
+	void gatherFiles(Set<File> allFiles) throws Exception {
```
Comment (bjhargrave) [inline biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java]: Why did this method loose `private`? I did not notice `gatherFiles` used anywhere in this PR.

**Comment by bjhargrave** [comment]
Overall, I am not sure of the value of this change. I made some specific comments on the code but I am not convinced this is a good idea. A repo should have some consistency in its layout and this includes file naming. Allowing any arbitrary file name rules defeats this value. So I don't think this is a great idea. But others should also provide their input.

File: biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java (hunk 1 above)
Comment (eli-jordan) [inline biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java]: That's fair, I didn't put a ton of thought into the names, I'll try to come up with some better names.

File: biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java (hunk 4 above)
Comment (eli-jordan) [inline biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java]: Yeah I thought about that, but the error appeared to be reported in bnd tools when I just let it bubble up. I can easily add exception handling here if that is best

File: biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java (hunk 3 above)
Comment (eli-jordan) [inline biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java]: No particular reason, easily changed.

File: biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java (hunk 5 above)
Comment (eli-jordan) [inline biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java]: Oops... My bad. Not super familiar with how marges work in egit

File: biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java (hunk 6 above)
Comment (eli-jordan) [inline biz.aQute.repository/src/aQute/bnd/deployer/repository/LocalIndexedRepo.java]: It's used in the tests I added, I call it directly to test that the files gathered respect the policy

**Comment by eli-jordan** [comment]
I think that's fair, but not all jar follow the strict naming convention required by the existing convention. I think there needs to be some more flexibility here. Perhaps allowing an arbitrary regular expression is making it too flexible. 

I look forward to further feedback!

**Comment by eli-jordan** [comment]
Ok, so I have come up with some better names for the properties etc.

The property key would be 'include' that takes a comma separated list of 
predefined patterns, and arbitrary regex is not allowed.

The predefined values would be;

'versionedJars' - include jars with the versioned bundle naming convention (the current implementation, and default if the 'include' prop is not set)
'jars' - include all jars i.e. regex .*jar$
'libs' - include all libs i.e. regex .*lib$
'bundle' - include any files that have a .jar suffix, where a manifest exists and has the header Bundle-MaifestVersion: 2
Note: I'm not convinced something like this is required, but would welcome feedback.
e.g.

...;include='jars,libs'

would include all files with the .jar suffix or the .lib suffix

Thoughts?

**Comment by eli-jordan** [comment]
I added what I described above without the ```bundle``` file inclusion policy

**Comment by pkriens** [comment]
It is interesting to pursue this a bit different.

First this should be a different repo, people might have assumptions about the fact this is is an old FileRepo. We could use Java 7 watching to pick changes as they go.

**Comment by eli-jordan** [comment]
I was very careful to ensure that the new behaviour is opt in. i.e. if the ```include``` parameter is not specified then the behaviour of this repo is exactly the same as it was prior to this change.

Using a java 7 file walker + watcher would definitely be a more performant way, and would not require manual refreshing. However, that is significantly more work than this simple change.

**Comment by pkriens** [comment]
Even if it is opt-in, I am worried that this plugin is a File Repo. I really think this should be a completely new repo.

Anyway, we're discussing some heavy changes to the repo model so I want to postpone this anyway until we know what we're going to do there.

**Comment by eli-jordan** [comment]
Ok fair enough. My need for this change was about indexing the bundles in a WAS Liberty Profile installation for use during development. The current LocalIndexedRepo doesn't index the bundles because they do not meet the naming criteria.

Would a much more modest change of the regex from

``` ([-a-zA-z0-9_\\.]+)-([0-9\\.]+)(-[-a-zA-z0-9_]+)?\\.(jar|lib) ```
to
``` ([-a-zA-z0-9_\\.]+)(-|_)([0-9\\.]+)(-[-a-zA-z0-9_]+)?\\.(jar|lib) ```

be accepted? 

This would address my use case.

**Comment by bjhargrave** [comment]
So this change is just to allow underscore as well as hyphen between the
name and version?  If so, then I am fine with that change.

On Mon, Nov 2, 2015, 12:36 eli-jordan <notifications@github.com> wrote:

> Ok fair enough. My need for this change was about indexing the bundles in
> a WAS Liberty Profile installation for use during development. The current
> LocalIndexedRepo doesn't index the bundles because they do not meet the
> naming criteria.
>
> Would a much more modest change of the regex from
>
> ([-a-zA-z0-9_\\.]+)-([0-9\\.]+)(-[-a-zA-z0-9_]+)?\\.(jar|lib)
>
> to
> ([-a-zA-z0-9_\\.]+)(-|_)([0-9\\.]+)(-[-a-zA-z0-9_]+)?\\.(jar|lib)
>
> be accepted?
>
> This would address my use case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bndtools/bnd/pull/1155#issuecomment-152994892>.
>
-- 
-- 
BJ

**Comment by bjhargrave** [comment]
So we should close this PR since #1178 has been merged.

**Pull request closed without merging**